src/openrct2-ui/windows/MapGen.cpp

```diff
@@ -1351,41 +1474,74 @@ namespace OpenRCT2::Ui::Windows
-                    return SimplexDraw(dpi);
-                case WINDOW_MAPGEN_PAGE_HEIGHTMAP:
-                    return HeightmapDraw(dpi);
+                case WINDOW_MAPGEN_PAGE_FORESTS:
+                    return ForestsDraw(dpi);
+                case WINDOW_MAPGEN_PAGE_TERRAIN:
+                    return TerrainDraw(dpi);
+                case WINDOW_MAPGEN_PAGE_WATER:
+                    return WaterDraw(dpi);
             }
         }
 
         void OnTextInput(WidgetIndex widgetIndex, std::string_view text) override
         {
+            auto strText = std::string(text);
+            char* end;
+
+            // Convert text to integer value
+            int32_t value{};
+            if (page == WINDOW_MAPGEN_PAGE_BASE && widgetIndex == WIDX_SIMPLEX_BASE_FREQ)
+                value = 100 * strtof(strText.c_str(), &end);
+            else
+                value = strtol(strText.c_str(), &end, 10);
+
+            if (*end != '\0')
+                return;
+
+            // Take care of unit conversion
+            int32_t rawValue = value;
+            if (page != WINDOW_MAPGEN_PAGE_BASE)
+            {
+                switch (Config::Get().general.MeasurementFormat)
+                {
+                    case MeasurementFormat::Imperial:
+                        value = FeetToMetres(value);
+                        [[fallthrough]];
+
+                    default:
+                        value = std::clamp(MetresToBaseZ(value), kMinimumLandHeight, kMaximumLandHeight);
+                        break;
+                }
+            }
+
+            // Pass on to the actual properties
             switch (page)
             {
                 case WINDOW_MAPGEN_PAGE_BASE:
-                    return TextInput(widgetIndex, text);
-                case WINDOW_MAPGEN_PAGE_SIMPLEX:
-                    return TextInput(widgetIndex, text);
+                    return BaseTextInput(widgetIndex, value);
+                case WINDOW_MAPGEN_PAGE_TERRAIN:
+                    return TerrainTextInput(widgetIndex, value);
+                case WINDOW_MAPGEN_PAGE_WATER:
+                    return WaterTextInput(widgetIndex, value);
+                case WINDOW_MAPGEN_PAGE_FORESTS:
+                    return ForestsTextInput(widgetIndex, rawValue, value);
             }
         }
 
         void AfterLoadingHeightMap(int32_t result, const utf8* path)
         {
             if (result == MODAL_RESULT_OK)
             {
-                if (!MapGenLoadHeightmap(path))
+                if (!MapGenLoadHeightmapImage(path))
                 {
                     // TODO: Display error popup
                     return;
                 }
 
                 // The window needs to be open while using the map
                 _heightmapLoaded = true;
-                SetPage(WINDOW_MAPGEN_PAGE_HEIGHTMAP);
-
-                HeightmapGenerateMap();
+                _heightmapFilename = std::string(fs::u8path(path).filename());
```

Comment:
Our Windows CI is complaining about this line. I added the explicit `std::string` constructor call in an attempt to mitigate this, but clearly to no avail. Any ideas?

Reply:
```suggestion
                _heightmapFilename = fs::u8path(path).filename().string();
```
Should do the trick